File: Data Structures & Algorithms/buy-and-sell-crypto/submission-0.cpp
class Solution {
public:
    int maxProfit(vector<int>& prices) {
        //I have two pointers, starts at idx 0 and 1, 
        //I compare val[1] to val[0], if val[1] is bigger or equal, I see if
        //the profit gives max profit (and save)
        //else if val[1] is smaller, I can discard val[0] as I know it will not 
        //be the best buy point and move my new left pter to val[1]
        int res = 0;
        int left = 0;
        int right = 1;
        while(right<prices.size()){
            if(prices[right]>=prices[left]){
                int prof = prices[right]-prices[left];
                res = max(res, prof);
            }
            else //pr<pl
            {
                left = right;
            }
            right++;

        }
        return res;
    }
};

File: Data Structures & Algorithms/buy-and-sell-crypto-with-cooldown/submission-0.py
class Solution:
    def maxProfit(self, prices: List[int]) -> int:
        #dp[i][x]: max profit day i to end, with x (0 or 1) coin currently
        #if i==0: return 0
        end = len(prices)-1
        # no coins
        #if x ==0: max(dp[i-1][1], dp[i-1][0] : buy or not buy
        #has coins
        #if x==1: max(sell+dp[i-2][0], dp[i-1][1])
        memo = {}
        def dp(i, x):
            if i>end:
                return 0
            if (i,x) in memo:
                return memo[(i,x)]
            if x==0:
                memo[(i,x)]=max(-prices[i]+dp(i+1,1), dp(i+1,0))
            if x==1:
                memo[(i,x)]=max(prices[i]+dp(i+2,0), dp(i+1,1))
            return memo[(i,x)]
        return dp(0,0)
